benchmark/src/main/scala/fs2/benchmark/CatenableBenchmark.scala

```diff
@@ -0,0 +1,36 @@
+package fs2
+package benchmark
+
+import org.openjdk.jmh.annotations.{Benchmark, State, Scope}
+
+import fs2.util.Catenable
+
+@State(Scope.Thread)
+class CatenableBenchmark extends BenchmarkUtils {
+
+  val smallCatenable = Catenable(1, 2, 3, 4, 5)
+  val smallVector = Vector(1, 2, 3, 4, 5)
+
+  val largeCatenable = Catenable.fromSeq(0 to 1000000)
+  val largeVector = (0 to 1000000).toVector
+
+  @Benchmark def mapSmallCatenable = smallCatenable.map(_ + 1)
+  @Benchmark def mapSmallVector = smallVector.map(_ + 1)
+  @Benchmark def mapLargeCatenable = largeCatenable.map(_ + 1)
+  @Benchmark def mapLargeVector = largeVector.map(_ + 1)
+
+  @Benchmark def foldLeftSmallCatenable = smallCatenable.foldLeft(0)(_ + _)
+  @Benchmark def foldLeftSmallVector = smallVector.foldLeft(0)(_ + _)
+  @Benchmark def foldLeftLargeCatenable = largeCatenable.foldLeft(0)(_ + _)
+  @Benchmark def foldLeftLargeVector = largeVector.foldLeft(0)(_ + _)
+
+  @Benchmark def consSmallCatenable = 0 +: smallCatenable
+  @Benchmark def consSmallVector = 0 +: smallVector
+  @Benchmark def consLargeCatenable = 0 +: largeCatenable
+  @Benchmark def consLargeVector = 0 +: largeVector
+
+  @Benchmark def createTinyCatenable = Catenable(1)
+  @Benchmark def createTinyVector = Vector(1)
+  @Benchmark def createSmallCatenable = Catenable(1, 2, 3, 4, 5)
```

Comment:
@pchiusano @edmundnoble I've done a lot of work with amortized factors in functional data structures (a few years ago).  The problem with amortization arguments in all contexts, but *especially* functional data structures, is they often require enormously high values of *n* to come into play.  Additionally, amortization arguments often ignore constant factors (in fact, that's literally what the argument is doing in an inductive fashion), which is not a fair argument to make when your constant factors are extremely high.

There are two excellent and classic examples of this problem: BankersQueue and FingerTree.  BankersQueue is basically just a lazy version of the eager "front/back list queue" thing that everyone's tried at least once.  And there is a *proof* that it is more efficient than the eager version… for some arbitrarily large value of *n*.  It turns out that, if you benchmark it, the eager version is almost twice as fast (in Scala) for any queue size that fits into 32bit memory, which is sort of insane.  FingerTree is a similar, even more dramatic example.  FingerTree is an amortized constant *double-ended queue*, which is something that the eager banker's queue can't provide, and so in a very real sense it is offering (and proving) better *asymptotes*, not just better amortization costs.  But on the JVM, and for queue sizes less than the billions, it is massively, *hilariously* slower than the alternatives.

So we have to be careful about this stuff.  Amortization arguments, discard by definition performance factors which are relevant and even dominant in practical workloads.  And this is especially true on the JVM and with a functional style, where amortization often relies in thunking and other pipeline- and PIC-defeating constructs that deoptimize code (by a constant factor) in ways below the level of algorithmic analysis.